python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1311,3 +1318,109 @@ class TableMetadataSet(NamespacedMetadataSet):
     @classmethod
     def namespace(cls) -> str:
         return "dagster"
+
+
+DEFAULT_SOURCE_FILE_KEY = "asset_definition"
+
+
+## Source Metadata
+
+
+@experimental
+@whitelist_for_serdes
+class LocalFileSource(DagsterModel):
+    """Represents a local file source location."""
+
+    file_path: str
+    line_number: int
+
+
+@experimental
+@whitelist_for_serdes
+class SourceMetadataValue(DagsterModel, MetadataValue["SourceMetadataValue"]):
+    """Metadata value type which represents source locations (locally or otherwise)
+    of the asset in question. For example, the file path and line number where the
+    asset is defined.
+
+    Properties:
+    sources (Dict[str, LocalFileSource]): A labeled dictionary of sources.
+    """
+
+    sources: Dict[str, LocalFileSource] = {}
+
+    @property
+    def value(self) -> "SourceMetadataValue":
+        return self
+
+
+def source_path_from_fn(fn: Callable[..., Any]) -> Optional[LocalFileSource]:
+    cwd = os.getcwd()
+    origin_file: Optional[str] = None
+    origin_line = None
+    try:
+        origin_file = os.path.abspath(os.path.join(cwd, inspect.getsourcefile(fn)))  # type: ignore
+        origin_file = check.not_none(origin_file)
+        origin_line = inspect.getsourcelines(fn)[1]
+    except TypeError:
+        return None
+
+    return LocalFileSource(
+        file_path=origin_file,
+        line_number=origin_line,
+    )
+
+
+class SourceDataMetadataSet(NamespacedMetadataSet):
+    """Metadata entries that apply to asset definitions and which specify the source code location
+    for the asset.
+    """
+
+    source_data: SourceMetadataValue
+
+    @classmethod
+    def namespace(cls) -> str:
+        return "dagster"
+
+
+def _with_code_source_single_definition(
+    assets_def: Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"],
+) -> Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"]:
+    from dagster._core.definitions.assets import AssetsDefinition
+
+    if not isinstance(assets_def, AssetsDefinition):
```

Comment:
This is because the other types don't have associated ops? `CacheableAssetsDefinition` has one, right?

Whatever it is, worth explaining it in a comment.

Reply:
Will handle cachable assets in stacked PR, it's a bit trickier there